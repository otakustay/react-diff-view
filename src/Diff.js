import {Children, cloneElement} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Hunk from './Hunk';
import {viewTypePropType, hunkPropType} from './propTypes';
import './Diff.css';

const Diff = ({diffType, hunks, children, className, ...props}) => {
    const monotonous = diffType === 'add' || diffType === 'delete';
    const cols = ((viewType, monotonous) => {
        if (viewType === 'unified') {
            return (
                <colgroup>
                    <col className="diff-gutter-col" />
                    <col className="diff-gutter-col" />
                    <col />
                </colgroup>
            );
        }

        if (monotonous) {
            return (
                <colgroup>
                    <col className="diff-gutter-col" />
                    <col />
                </colgroup>
            );
        }

        return (
            <colgroup>
                <col className="diff-gutter-col" />
                <col />
                <col className="diff-gutter-col" />
                <col />
            </colgroup>
        );
    })(props.viewType, monotonous);
    const hunksChildren = children
        ? Children.map(children, child => cloneElement(child, {...props, monotonous}))
        : hunks.map(hunk => <Hunk key={hunk.content} hunk={hunk} monotonous={monotonous} {...props} />);

    return (
        <table className={classNames('diff', className)}>
            {cols}
            {hunksChildren}
        </table>
    );
};

Diff.propTypes = {
    // TODO: [2.x] `diffType` should be required.
    diffType: PropTypes.oneOf(['add', 'delete', 'modify', 'rename', 'copy']),
    viewType: viewTypePropType.isRequired,
    hunks: PropTypes.arrayOf(hunkPropType),
    gutterAnchor: PropTypes.bool,
    generateAnchorID: PropTypes.func,
    children: PropTypes.oneOfType([PropTypes.element, PropTypes.arrayOf(PropTypes.element)])
};

Diff.defaultProps = {
    diffType: 'modify',
    hunks: undefined,
    children: undefined,
    gutterAnchor: false,
    generateAnchorID() {
        return undefined;
    }
};

export default Diff;
