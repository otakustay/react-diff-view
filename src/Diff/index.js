import {memo, useRef, useCallback} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {Provider} from '../context';
import Hunk from '../Hunk';
import './index.css';

const noop = () => {}; // eslint-disable-line no-empty-function

const findClosest = (target, className) => {
    while (target && target !== document && !target.classList.contains(className)) {
        target = target.parentNode; // eslint-disable-line no-param-reassign
    }

    return target === document ? null : target;
};

const setUserSelectStyle = (element, selectable) => {
    const value = selectable ? 'auto' : 'none';

    if (element.style.userSelect !== value) {
        element.style.userSelect = value; // eslint-disable-line no-param-reassign
    }
};

const Diff = props => {
    const root = useRef(null);
    const enableColumnSelection = useCallback(
        ({target}) => {
            const closestCell = findClosest(target, 'diff-code');

            if (!closestCell) {
                return;
            }

            window.getSelection().removeAllRanges();

            const index = [...closestCell.parentNode.children].indexOf(closestCell);

            if (index !== 1 && index !== 3) {
                return;
            }

            const lines = root.current.querySelectorAll('.diff-line');
            for (const line of lines) {
                const cells = line.children;
                setUserSelectStyle(cells[1], index === 1);
                setUserSelectStyle(cells[3], index === 3);
            }
        },
        []
    );

    const {diffType, children, className, optimizeSelection, hunks, ...remainings} = props;
    const {gutterType, viewType} = remainings;
    const hideGutter = gutterType === 'none';
    const monotonous = diffType === 'add' || diffType === 'delete';
    const onTableMouseDown = (viewType === 'split' && !monotonous && optimizeSelection) ? enableColumnSelection : noop;
    const cols = ((viewType, monotonous) => {
        if (viewType === 'unified') {
            return (
                <colgroup>
                    {!hideGutter && <col className="diff-gutter-col" />}
                    {!hideGutter && <col className="diff-gutter-col" />}
                    <col />
                </colgroup>
            );
        }

        if (monotonous) {
            return (
                <colgroup>
                    {!hideGutter && <col className="diff-gutter-col" />}
                    <col />
                </colgroup>
            );
        }

        return (
            <colgroup>
                {!hideGutter && <col className="diff-gutter-col" />}
                <col />
                {!hideGutter && <col className="diff-gutter-col" />}
                <col />
            </colgroup>
        );
    })(props.viewType, monotonous);

    return (
        <Provider value={{...remainings, monotonous}}>
            <table
                ref={root}
                className={classNames('diff', `diff-${viewType}`, className)}
                onMouseDown={onTableMouseDown}
            >
                {cols}
                {children(hunks)}
            </table>
        </Provider>
    );
};

Diff.propTypes = {
    diffType: PropTypes.oneOf(['add', 'delete', 'modify', 'rename', 'copy']).isRequired,
    viewType: PropTypes.oneOf(['unified', 'split']).isRequired,
    hunks: PropTypes.arrayOf(PropTypes.object).isRequired,
    gutterType: PropTypes.oneOf(['default', 'none', 'anchor']),
    generateAnchorID: PropTypes.func,
    selectedChanges: PropTypes.arrayOf(PropTypes.string),
    widgets: PropTypes.objectOf(PropTypes.node),
    optimizeSelection: PropTypes.bool,
    className: PropTypes.string,
    renderToken: PropTypes.func,
    children: PropTypes.func,
};

Diff.defaultProps = {
    gutterType: 'default',
    optimizeSelection: false,
    selectedChanges: [],
    widgets: {},
    className: '',
    renderToken: undefined,
    generateAnchorID() {
        return undefined;
    },
    children(hunks) {
        const key = hunk => `-${hunk.oldStart},${hunk.oldLines} +${hunk.newStart},${hunk.newLines}`;
        return hunks.map(hunk => <Hunk key={key(hunk)} hunk={hunk} />);
    },
};

export default memo(Diff);
