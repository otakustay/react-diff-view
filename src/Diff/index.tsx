import {memo, useRef, useCallback, ReactElement, MouseEvent, useMemo} from 'react';
import classNames from 'classnames';
import {
    ContextProps,
    EventMap,
    GutterType,
    HunkTokens,
    Provider,
    ViewType,
    RenderToken,
    RenderGutter,
    DEFAULT_CONTEXT_VALUE,
} from '../context';
import Hunk from '../Hunk';
import {ChangeData, HunkData} from '../utils';

export type DiffType = 'add' | 'delete' | 'modify' | 'rename' | 'copy';

export interface DiffProps {
    diffType: DiffType;
    hunks: HunkData[];
    viewType?: ViewType;
    gutterType?: GutterType;
    generateAnchorID?: (change: ChangeData) => string | undefined;
    selectedChanges?: string[];
    widgets?: Record<string, ReactElement>;
    optimizeSelection?: boolean;
    className?: string;
    hunkClassName?: string;
    lineClassName?: string;
    gutterClassName?: string;
    codeClassName?: string;
    tokens?: HunkTokens;
    renderToken?: RenderToken;
    renderGutter?: RenderGutter;
    gutterEvents?: EventMap;
    codeEvents?: EventMap;
    children?: (hunks: HunkData[]) => ReactElement | ReactElement[];
}

function noop() {}

function findClosest(target: HTMLElement, className: string) {
    let current: HTMLElement | null = target;
    while (current && current !== document.documentElement && !current.classList.contains(className)) {
        current = current.parentElement;
    }

    return current === document.documentElement ? null : current;
}

function setUserSelectStyle(element: Element, selectable: boolean) {
    const value = selectable ? 'auto' : 'none';

    if (element instanceof HTMLElement && element.style.userSelect !== value) {
        element.style.userSelect = value; // eslint-disable-line no-param-reassign
    }
}

function defaultRenderChildren(hunks: HunkData[]) {
    const key = (hunk: HunkData) => `-${hunk.oldStart},${hunk.oldLines} +${hunk.newStart},${hunk.newLines}`;
    return hunks.map(hunk => <Hunk key={key(hunk)} hunk={hunk} />);
}

function Diff(props: DiffProps) {
    const {
        diffType,
        hunks,
        optimizeSelection,
        className,
        hunkClassName = DEFAULT_CONTEXT_VALUE.hunkClassName,
        lineClassName = DEFAULT_CONTEXT_VALUE.lineClassName,
        gutterClassName = DEFAULT_CONTEXT_VALUE.gutterClassName,
        codeClassName = DEFAULT_CONTEXT_VALUE.codeClassName,
        gutterType = DEFAULT_CONTEXT_VALUE.gutterType,
        viewType = DEFAULT_CONTEXT_VALUE.viewType,
        gutterEvents = DEFAULT_CONTEXT_VALUE.gutterEvents,
        codeEvents = DEFAULT_CONTEXT_VALUE.codeEvents,
        generateAnchorID = DEFAULT_CONTEXT_VALUE.generateAnchorID,
        selectedChanges = DEFAULT_CONTEXT_VALUE.selectedChanges,
        widgets = DEFAULT_CONTEXT_VALUE.widgets,
        renderGutter = DEFAULT_CONTEXT_VALUE.renderGutter,
        tokens,
        renderToken,
        children = defaultRenderChildren,
    } = props;
    const root = useRef<HTMLTableElement>(null);
    const enableColumnSelection = useCallback(
        ({target, button}: MouseEvent<HTMLElement>) => {
            if (button !== 0) {
                return;
            }

            const closestCell = findClosest(target as HTMLElement, 'diff-code');

            if (!closestCell || !closestCell.parentElement) {
                return;
            }

            const selection = window.getSelection();
            if (selection) {
                selection.removeAllRanges();
            }

            const index = [...closestCell.parentElement.children].indexOf(closestCell);

            if (index !== 1 && index !== 3) {
                return;
            }

            const lines = root.current ? root.current.querySelectorAll('.diff-line') : [];
            for (const line of lines) {
                const cells = line.children;
                setUserSelectStyle(cells[1], index === 1);
                setUserSelectStyle(cells[3], index === 3);
            }
        },
        []
    );
    const hideGutter = gutterType === 'none';
    const monotonous = diffType === 'add' || diffType === 'delete';
    const onTableMouseDown = (viewType === 'split' && !monotonous && optimizeSelection) ? enableColumnSelection : noop;
    const cols = useMemo(
        () => {
            if (viewType === 'unified') {
                return (
                    <colgroup>
                        {!hideGutter && <col className="diff-gutter-col" />}
                        {!hideGutter && <col className="diff-gutter-col" />}
                        <col />
                    </colgroup>
                );
            }

            if (monotonous) {
                return (
                    <colgroup>
                        {!hideGutter && <col className="diff-gutter-col" />}
                        <col />
                    </colgroup>
                );
            }

            return (
                <colgroup>
                    {!hideGutter && <col className="diff-gutter-col" />}
                    <col />
                    {!hideGutter && <col className="diff-gutter-col" />}
                    <col />
                </colgroup>
            );
        },
        [viewType, monotonous, hideGutter]
    );
    // TODO: in later versions, we can split context into multiple to reduce component render
    const settingsContextValue = useMemo(
        (): ContextProps => {
            return {
                hunkClassName,
                lineClassName,
                gutterClassName,
                codeClassName,
                monotonous,
                hideGutter,
                viewType,
                gutterType,
                codeEvents,
                gutterEvents,
                generateAnchorID,
                selectedChanges,
                widgets,
                renderGutter,
                tokens,
                renderToken,
            };
        },
        [
            codeClassName,
            codeEvents,
            generateAnchorID,
            gutterClassName,
            gutterEvents,
            gutterType,
            hideGutter,
            hunkClassName,
            lineClassName,
            monotonous,
            renderGutter,
            renderToken,
            selectedChanges,
            tokens,
            viewType,
            widgets,
        ]
    );

    return (
        <Provider value={settingsContextValue}>
            <table
                ref={root}
                className={classNames('diff', `diff-${viewType}`, className)}
                onMouseDown={onTableMouseDown}
            >
                {cols}
                {children(hunks)}
            </table>
        </Provider>
    );
}

export default memo(Diff);
