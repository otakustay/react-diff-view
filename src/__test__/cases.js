/* In case you are looking for type of hunk, see src/Hunk/index.js */
import {parseDiff} from '..';

export const basic = `diff --git a/src/__test__/index.test.jsx b/src/__test__/index.test.jsx
index 643c2f0..7883597 100644
--- a/src/__test__/index.test.jsx
+++ b/src/__test__/index.test.jsx
@@ -21,3 +21,3 @@ describe('basic test', () => {
     test('App renders correctly', () => {
-        expect(renderer.create(<App diffText={'deff'} />).toJSON()).toMatchSnapshot();
+        expect(renderer.create(<App diffText={'diff'} />).toJSON()).toMatchSnapshot();
     });`;

export const basicHunk = parseDiff(basic)[0].hunks[0];

export const multiple = `diff --git a/src/index1.js b/src/index1.js
index e69de29..d00491f 100644
--- a/src/index1.js
+++ b/src/index1.js
@@ -0,0 +1 @@
+1
diff --git a/src/index.js b/src/index2.js
index d00491f..0cfbf08 100644
--- a/src/index2.js
+++ b/src/index2.js
@@ -1 +1 @@
-1
+2
diff --git a/src/index3.js b/src/index3.js
index 0cfbf08..e69de29 100644
--- a/src/index3.js
+++ b/src/index3.js
@@ -1 +0,0 @@
-2`;

export const multipleHunk = parseDiff(basic)[0].hunks[0];
