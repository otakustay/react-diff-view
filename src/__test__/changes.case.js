
export const normalChange = {isNormal: true, lineNumber: 0, oldLineNumber: 0};
export const insertChange = {isInsert: true, lineNumber: 0, oldLineNumber: 0};
export const deleteChange = {isDelete: true, lineNumber: 0, oldLineNumber: 0};
