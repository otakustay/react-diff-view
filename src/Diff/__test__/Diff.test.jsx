import renderer from 'react-test-renderer';
import Diff from '..';
import {Decoration, getChangeKey} from '../..';
import {basicHunk} from '../../__test__/cases';

const DiffSplit = ({children}) => (
    <Diff diffType={'modify'} hunks={[basicHunk]} viewType={'split'}>
        {children}
    </Diff>
);

const DiffUnified = ({children}) => (
    <Diff diffType={'modify'} hunks={[basicHunk]} viewType={'unified'}>
        {children}
    </Diff>
);

// TODO it's weird that children should be a function
describe('Diff', () => {
    const renderRawHunks = hunks => <div>{JSON.stringify(hunks)}</div>;

    test('renders correctly', () => {
        expect(renderer.create(<DiffSplit>{renderRawHunks}</DiffSplit>).toJSON()).toMatchSnapshot();
    });

    test('unified Diff', () => {
        expect(renderer.create(<DiffUnified>{renderRawHunks}</DiffUnified>).toJSON()).toMatchSnapshot();
    });
});

describe('Diff with Decoration', () => {
    const renderDecoration = () => <Decoration><div>xxx</div></Decoration>;

    test('renders correctly', () => {
        expect(renderer.create(<DiffSplit>{renderDecoration}</DiffSplit>).toJSON()).toMatchSnapshot();
    });

    test('unified Diff with Decoration', () => {
        expect(renderer.create(<DiffUnified>{renderDecoration}</DiffUnified>).toJSON()).toMatchSnapshot();
    });
});

const getWidgets = hunks => {
    const changes = hunks.reduce((result, {changes}) => [...result, ...changes], []);
    const longLines = changes.filter(({content}) => content.length > 20);
    return longLines.reduce(
        (widgets, change) => {
            const changeKey = getChangeKey(change);

            return {
                ...widgets,
                [changeKey]: <span className="error">Line too long</span>
            };
        },
        {}
    );
};

describe('Diff with Widget', () => {
    test('split widget', () => {
        const hunks = [basicHunk];
        expect(renderer.create(
            <Diff hunks={hunks} widgets={getWidgets(hunks)} diffType="modify" viewType="split" />
        ).toJSON()).toMatchSnapshot();
    });

    test('unified widget', () => {
        const hunks = [basicHunk];
        expect(renderer.create(
            <Diff hunks={hunks} widgets={getWidgets(hunks)} diffType="modify" viewType="unified" />
        ).toJSON()).toMatchSnapshot();
    });
});
