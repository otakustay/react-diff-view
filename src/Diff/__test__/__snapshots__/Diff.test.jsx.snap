// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Diff renders correctly 1`] = `
<table
  className="diff diff-split"
  onMouseDown={[Function]}
>
  <colgroup>
    <col
      className="diff-gutter-col"
    />
    <col />
    <col
      className="diff-gutter-col"
    />
    <col />
  </colgroup>
  <div>
    [{"content":"@@ -21,3 +21,3 @@ describe('basic test', () =&gt; {","oldStart":21,"newStart":21,"oldLines":3,"newLines":3,"changes":[{"content":"    test('App renders correctly', () =&gt; {","type":"normal","isNormal":true,"oldLineNumber":21,"newLineNumber":21},{"content":"        expect(renderer.create(&lt;App diffText={'deff'} /&gt;).toJSON()).toMatchSnapshot();","type":"delete","isDelete":true,"lineNumber":22},{"content":"        expect(renderer.create(&lt;App diffText={'diff'} /&gt;).toJSON()).toMatchSnapshot();","type":"insert","isInsert":true,"lineNumber":22},{"content":"    });","type":"normal","isNormal":true,"oldLineNumber":23,"newLineNumber":23}],"isPlain":false}]
  </div>
</table>
`;

exports[`Diff unified Diff 1`] = `
<table
  className="diff diff-unified"
  onMouseDown={[Function]}
>
  <colgroup>
    <col
      className="diff-gutter-col"
    />
    <col
      className="diff-gutter-col"
    />
    <col />
  </colgroup>
  <div>
    [{"content":"@@ -21,3 +21,3 @@ describe('basic test', () =&gt; {","oldStart":21,"newStart":21,"oldLines":3,"newLines":3,"changes":[{"content":"    test('App renders correctly', () =&gt; {","type":"normal","isNormal":true,"oldLineNumber":21,"newLineNumber":21},{"content":"        expect(renderer.create(&lt;App diffText={'deff'} /&gt;).toJSON()).toMatchSnapshot();","type":"delete","isDelete":true,"lineNumber":22},{"content":"        expect(renderer.create(&lt;App diffText={'diff'} /&gt;).toJSON()).toMatchSnapshot();","type":"insert","isInsert":true,"lineNumber":22},{"content":"    });","type":"normal","isNormal":true,"oldLineNumber":23,"newLineNumber":23}],"isPlain":false}]
  </div>
</table>
`;

exports[`Diff with Decoration renders correctly 1`] = `
<table
  className="diff diff-split"
  onMouseDown={[Function]}
>
  <colgroup>
    <col
      className="diff-gutter-col"
    />
    <col />
    <col
      className="diff-gutter-col"
    />
    <col />
  </colgroup>
  <tbody
    className="diff-decoration"
  >
    <tr>
      <td
        className="diff-decoration-content"
        colSpan={4}
      >
        <div>
          xxx
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Diff with Decoration unified Diff with Decoration 1`] = `
<table
  className="diff diff-unified"
  onMouseDown={[Function]}
>
  <colgroup>
    <col
      className="diff-gutter-col"
    />
    <col
      className="diff-gutter-col"
    />
    <col />
  </colgroup>
  <tbody
    className="diff-decoration"
  >
    <tr>
      <td
        className="diff-decoration-content"
        colSpan={3}
      >
        <div>
          xxx
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Diff with Widget split widget 1`] = `
<table
  className="diff diff-split"
  onMouseDown={[Function]}
>
  <colgroup>
    <col
      className="diff-gutter-col"
    />
    <col />
    <col
      className="diff-gutter-col"
    />
    <col />
  </colgroup>
  <tbody
    className="diff-hunk"
  >
    <tr
      className="diff-line diff-line-normal"
    >
      <td
        className="diff-gutter diff-gutter-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        21
      </td>
      <td
        className="diff-code diff-code-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
            test('App renders correctly', () =&gt; {
      </td>
      <td
        className="diff-gutter diff-gutter-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        21
      </td>
      <td
        className="diff-code diff-code-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
            test('App renders correctly', () =&gt; {
      </td>
    </tr>
    <tr
      className="diff-widget"
    >
      <td
        className="diff-widget-content"
        colSpan={4}
      >
        <span
          className="error"
        >
          Line too long
        </span>
      </td>
    </tr>
    <tr
      className="diff-line diff-line-compare"
    >
      <td
        className="diff-gutter diff-gutter-delete"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        22
      </td>
      <td
        className="diff-code diff-code-delete"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
                expect(renderer.create(&lt;App diffText={'deff'} /&gt;).toJSON()).toMatchSnapshot();
      </td>
      <td
        className="diff-gutter diff-gutter-insert"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        22
      </td>
      <td
        className="diff-code diff-code-insert"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
                expect(renderer.create(&lt;App diffText={'diff'} /&gt;).toJSON()).toMatchSnapshot();
      </td>
    </tr>
    <tr
      className="diff-widget"
    >
      <td
        className="diff-widget-content"
        colSpan={2}
      >
        <span
          className="error"
        >
          Line too long
        </span>
      </td>
      <td
        className="diff-widget-content"
        colSpan={2}
      >
        <span
          className="error"
        >
          Line too long
        </span>
      </td>
    </tr>
    <tr
      className="diff-line diff-line-normal"
    >
      <td
        className="diff-gutter diff-gutter-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        23
      </td>
      <td
        className="diff-code diff-code-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
            });
      </td>
      <td
        className="diff-gutter diff-gutter-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        23
      </td>
      <td
        className="diff-code diff-code-normal"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
            });
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Diff with Widget unified widget 1`] = `
<table
  className="diff diff-unified"
  onMouseDown={[Function]}
>
  <colgroup>
    <col
      className="diff-gutter-col"
    />
    <col
      className="diff-gutter-col"
    />
    <col />
  </colgroup>
  <tbody
    className="diff-hunk"
  >
    <tr
      className="diff-line"
    >
      <td
        className="diff-gutter diff-gutter-normal"
      >
        21
      </td>
      <td
        className="diff-gutter diff-gutter-normal"
      >
        21
      </td>
      <td
        className="diff-code diff-code-normal"
      >
            test('App renders correctly', () =&gt; {
      </td>
    </tr>
    <tr
      className="diff-widget"
    >
      <td
        className="diff-widget-content"
        colSpan={3}
      >
        <span
          className="error"
        >
          Line too long
        </span>
      </td>
    </tr>
    <tr
      className="diff-line"
    >
      <td
        className="diff-gutter diff-gutter-delete"
      >
        22
      </td>
      <td
        className="diff-gutter diff-gutter-delete"
      />
      <td
        className="diff-code diff-code-delete"
      >
                expect(renderer.create(&lt;App diffText={'deff'} /&gt;).toJSON()).toMatchSnapshot();
      </td>
    </tr>
    <tr
      className="diff-widget"
    >
      <td
        className="diff-widget-content"
        colSpan={3}
      >
        <span
          className="error"
        >
          Line too long
        </span>
      </td>
    </tr>
    <tr
      className="diff-line"
    >
      <td
        className="diff-gutter diff-gutter-insert"
      />
      <td
        className="diff-gutter diff-gutter-insert"
      >
        22
      </td>
      <td
        className="diff-code diff-code-insert"
      >
                expect(renderer.create(&lt;App diffText={'diff'} /&gt;).toJSON()).toMatchSnapshot();
      </td>
    </tr>
    <tr
      className="diff-widget"
    >
      <td
        className="diff-widget-content"
        colSpan={3}
      >
        <span
          className="error"
        >
          Line too long
        </span>
      </td>
    </tr>
    <tr
      className="diff-line"
    >
      <td
        className="diff-gutter diff-gutter-normal"
      >
        23
      </td>
      <td
        className="diff-gutter diff-gutter-normal"
      >
        23
      </td>
      <td
        className="diff-code diff-code-normal"
      >
            });
      </td>
    </tr>
  </tbody>
</table>
`;
